### PR TITLE
feat: generalize ConsensusDataProvider for manual-seal

### PR DESCRIPTION
**Pull request by yjhmelody**
It make manual-seal could handle proof.

Some implementation details can still be discussed here. 
I think the `append_block_import` could be redesign in some other ways for more useful.
e.g. handle proposal directly and return BlockImportParams or add more stages to handle.

### CODE REVIEW

**Comment by yjhmelody** [comment]
Anyone could give me a feedback?

File: client/consensus/manual-seal/src/consensus/aura.rs [hunk 1]
```diff
@@ -58,7 +58,7 @@ where
 	}
 }
 
-impl<B, C> ConsensusDataProvider<B> for AuraConsensusDataProvider<B, C>
+impl<B, C, PROOF> ConsensusDataProvider<B> for AuraConsensusDataProvider<B, C, PROOF>
```
Comment (bkchr) [inline client/consensus/manual-seal/src/consensus/aura.rs:61]:
```suggestion
impl<B, C, Proof> ConsensusDataProvider<B> for AuraConsensusDataProvider<B, C, Proof>
```

**Review by bkchr — COMMENTED** [review]
I'm not 100% sure about this, but it looks okay.

Please rename `PROOF` everywhere to `Proof`. Otherwise it looks okay.

**Review by davxy — COMMENTED** [review]
`<Pedantic Mode Nits ON>`
Can we stick to a naming convention for generic parameters?
In some places `P` has been used while in others `Proof` (or `PROOF`).
Generally (not everywhere and if there are no conflicts in names initial letters) we tend to use uppercase single letters or multiple letters for acronyms.
So (IMO) in this case `P` is preferred.
`<Pedantic Mode Nits OFF>`
:-)

**Comment by yjhmelody** [comment]
> `<Pedantic Mode Nits ON>` Can we stick to a naming convention for generic parameters? In some places `P` has been used while in others `Proof` (or `PROOF`). Generally (not everywhere and if there are no conflicts in names initial letters) we tend to use uppercase single letters or multiple letters for acronyms. So (IMO) in this case `P` is preferred. `<Pedantic Mode Nits OFF>` :-)

Ok, I agree with you.

**Comment by bkchr** [comment]
> `<Pedantic Mode Nits ON>` Can we stick to a naming convention for generic parameters? In some places `P` has been used while in others `Proof` (or `PROOF`). Generally (not everywhere and if there are no conflicts in names initial letters) we tend to use uppercase single letters or multiple letters for acronyms. So (IMO) in this case `P` is preferred. `<Pedantic Mode Nits OFF>` :-)

I mean we don't use `PROOF` anywhere, this was the reason why I don't wanted to have it. For me there is no fixed threshold, but when the number of generic parameters starts to explode, I switch to use better naming than single characters. 

It is okay and we can keep the pr as is now, but just some note for the future.

**Review by bkchr — APPROVED** [review]

**Review by davxy — CHANGES_REQUESTED** [review]
@yjhmelody I went through your PR trying to understand why is this required in the first place.

The only motivation that I see is that you have your custom implementation of `ConsensusDataProvider` trait and there you need to use somehow this new associated type `Proof` (e.g. to use your consensus instead of aura and babe).

But from what I can see, the `AuraConsensusDataProvider` and `BabeConsensusDataProvider` implementations doesn't use the introduced generic `P` nor the associated type `Proof` in any useful way. For them, you can just remove `P` and set `Proof` to `()`.

In the end, can you please explain why you need this and eventually change the `Aura` and `Babe` consensus data providers to something like:

```
impl<B, C> ConsensusDataProvider<B> for AuraConsensusDataProvider<C> {
    type Proof = ();
...
}

impl<B, C> ConsensusDataProvider<B> for BabeConsensusDataProvider<B, C> {
    type Proof = ();
...
}
```

**Comment by yjhmelody** [comment]
@davxy Hi, I want to mock some proof data in manual-seal mode. And maybe I use `aura`, so  `type Proof = ()` is not generic enogh, you know that `basic-authorship` could create `StorageProof`, I may use it.

File: client/consensus/manual-seal/src/consensus.rs [hunk 2]
```diff
@@ -33,6 +33,8 @@ pub trait ConsensusDataProvider<B: BlockT>: Send + Sync {
 	/// Block import transaction type
 	type Transaction;
 
+	type Proof: Send + Sync + 'static;
```
Comment (davxy) [inline client/consensus/manual-seal/src/consensus.rs:36]:
```suggestion
	/// The proof type.
	type Proof;
```
Bounds not strictly required here

File: client/consensus/manual-seal/src/consensus/aura.rs [hunk 3]
```diff
@@ -67,8 +67,10 @@ where
 		+ UsageProvider<B>
 		+ ProvideRuntimeApi<B>,
 	C::Api: AuraApi<B, AuthorityId>,
+	P: Send + Sync + 'static,
```
Comment (davxy) [inline client/consensus/manual-seal/src/consensus/aura.rs:70]:
```suggestion
	P: Send + Sync,
```
Lifetime bound not strictly required here

File: client/consensus/manual-seal/src/consensus/babe.rs [hunk 4]
```diff
@@ -190,8 +198,10 @@ where
 		+ UsageProvider<B>
 		+ ProvideRuntimeApi<B>,
 	C::Api: BabeApi<B>,
+	P: Send + Sync + 'static,
```
Comment (davxy) [inline client/consensus/manual-seal/src/consensus/babe.rs:201]:
```suggestion
	P: Send + Sync,
```
Lifetime bound not strictly required here

**Review by davxy — COMMENTED** [review]
Mmmh I see. So you want to use this additional generic to bind the struct to a specific proof type in case someone/somewhere want to use it.

It just looked a bit awkward because we're not currently actively using it anywhere (in our codebase).

Can't you directly implement the `ConsensusDataProvider` for your mock in order to prevent adding this extra generic param here?

I'm going to remove the blocker, but let's wait a bit for other opinions from @paritytech/sdk-node

**Comment by yjhmelody** [comment]
> Mmmh I see. So you want to use this additional generic to bind the struct to a specific proof type in case someone/somewhere want to use it.
> 
> 
> 
> It just looked a bit awkward because we're not currently actively using it anywhere (in our codebase).
> 
> 
> 
> Can't you directly implement the `ConsensusDataProvider` for your mock in order to prevent adding this extra generic param here?
> 
> 
> 
> I'm going to remove the blocker, but let's wait a bit for other opinions from @paritytech/sdk-node 
yeah，it is fine too

**Comment by koute** [comment]
> I'm going to remove the blocker, but let's wait a bit for other opinions from @paritytech/sdk-node

Honestly, while having so many generic parameters makes me think that this code should be ~~killed with fire~~ refactored somehow I think adding this seems reasonable, *if there's an actual use case that needs it*. We've already crossed the point of no return way earlier than the 7 generic parameters which one of the functions/types here accept, so one more's not going to make much difference.

**Review by davxy — APPROVED** [review]

**Comment by bkchr** [comment]
None of this code here is really "important". We also mainly just drag it with us. Meaning if someone needs something, we can add it. As long as the proposed changes are not that crazy or whatever.

**Pull request merged**